source/common/filesystem/filesystem_impl.cc

```diff
@@ -169,9 +171,20 @@ void FileImpl::flushThreadFunc() {
         stats_.reopen_failed_.inc();
       }
     }
+  }
+}
 
-    lock.lock();
+void FileImpl::flush() {
+  std::unique_lock<std::mutex> write_lock(write_lock_);
+  if (flush_buffer_.length() == 0) {
+    return;
   }
+
+  std::unique_lock<std::mutex> flush_buffer_lock(flush_lock_);
```

Comment:
Move the follow 4 lines into a helper and share w/ flushThreadFunc() above?

Reply:
I thought about that, but they're not quite common enough.  flushThreadFunc() has those lines, but interleaved with other things.  Also, doing write_lock.unlock() in another function makes the locking a bit harder to follow.

I could do the check-for-reopen to before that chunk and probably make it work, but I think it would complicate the locking and make it harder to follow.